src/protocol/libp2p/kademlia/bucket.rs

```diff
@@ -101,12 +101,10 @@ impl KBucket {
 
     /// Get iterator over the k-bucket, sorting the k-bucket entries in increasing order
     /// by distance.
-    pub fn closest_iter<'a, K: Clone>(
-        &'a mut self,
-        target: Key<K>,
-    ) -> impl Iterator<Item = &'a KademliaPeer> {
-        self.nodes.sort_by(|a, b| target.distance(&a.key).cmp(&target.distance(&b.key)));
-        self.nodes.iter().filter(|peer| !peer.addresses.is_empty())
+    pub fn closest_iter<K: Clone>(&self, target: &Key<K>) -> impl Iterator<Item = KademliaPeer> {
+        let mut nodes = self.nodes.clone();
```

Comment:
I'm not super happy about this `clone()` — a better way would be to only clone nodes that we finally yield from `RoutingTable::closest()`. But without this clone we'll need to manually iterate over buckets and peers within them, and the trick with nesting `ClosestBucketsIter` and `KBucket::closest_iter` and flattening them won't work due to borrowing of `self` in closures in `RoutingTable::closest()`.

Reply:
Maybe create an issue if it's worth revisiting at some point in the future? While I agree with you that it could be better, this PR is already a big improvement over the previous implementation so I'm fine with merging this as is